agent/dockerclient/dockerapi/docker_client.go

```diff
@@ -1244,24 +1245,53 @@ func (dg *dockerGoClient) Stats(id string, ctx context.Context) (<-chan *docker.
 		return nil, err
 	}
 
-	stats := make(chan *docker.Stats)
-	options := docker.StatsOptions{
-		ID:                id,
-		Stats:             stats,
-		Stream:            true,
-		Context:           ctx,
-		InactivityTimeout: StatsInactivityTimeout,
-	}
+	returnStats := make(chan *docker.Stats)
 
-	go func() {
-		statsErr := client.Stats(options)
-		if statsErr != nil {
-			seelog.Infof("DockerGoClient: Unable to retrieve stats for container %s: %v",
-				id, statsErr)
+	if !dg.config.PollMetrics {
+		options := docker.StatsOptions{
+			ID:                id,
+			Stats:             returnStats,
+			Stream:            true,
+			Context:           ctx,
+			InactivityTimeout: StatsInactivityTimeout,
 		}
-	}()
+	
+		go func() {
+			statsErr := client.Stats(options)
+			if statsErr != nil {
+				seelog.Infof("DockerGoClient: Unable to retrieve stats for container %s: %v",
+					id, statsErr)
+			}
+		}()
+	} else {
+
+		//Sleep for a random period time up to the polling interval. This will help make containers ask for stats at different times
+		time.Sleep(time.Second * time.Duration(rand.Intn(int(dg.config.PollingMetricsWaitDuration.Seconds()))))
+
+		go func() {
+			for {
+
+				stats := make(chan *docker.Stats, 1)
+				options := docker.StatsOptions{
+					ID:                id,
+					Stats:             stats,
+					Stream:            false,
+					Context:           ctx,
+					InactivityTimeout: StatsInactivityTimeout,
+				}
+				client.Stats(options)
+
+				dockerStats, ok := <-stats
+				if ok {
+					returnStats <- dockerStats
+				}
+
+				time.Sleep(dg.config.PollingMetricsWaitDuration)
```

Comment:
Instead of using `sleep`, you can use a `timer` here.

Reply:
Converted to use timer